libraries/c_sdk/standard/ble/src/iot_ble_mqtt_transport.c

```diff
@@ -0,0 +1,1081 @@
+
+    bufToSend.pBuffer = sharedBuffer;
+    bufToSend.size = RECV_PACKET_BUFFER_SIZE;
+
+    serializerInitVal = IotBleMqtt_InitSerialize();
+
+    if( serializerInitVal == true )
+    {
+        switch( packetType )
+        {
+            case MQTT_PACKET_TYPE_CONNECT:
+                status = handleOutgoingConnect( buf, &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_PUBLISH:
+                status = handleOutgoingPublish( buf, &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_PUBACK:
+                status = handleOutgoingPuback( buf, &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_SUBSCRIBE:
+                status = handleOutgoingSubscribe( buf, &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_UNSUBSCRIBE:
+                status = handleOutgoingUnsubscribe( buf, &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_PINGREQ:
+                status = handleOutgoingPingReq( &bufToSend, &pSize );
+                break;
+
+            case MQTT_PACKET_TYPE_DISCONNECT:
+                status = handleOutgoingDisconnect( &bufToSend, &pSize );
+                break;
+
+            /* QoS 2 cases, currently not supported by BLE */
+            case MQTT_PACKET_TYPE_PUBREC:
+            case MQTT_PACKET_TYPE_PUBREL:
+            case MQTT_PACKET_TYPE_PUBCOMP:
+                status = MQTTSendFailed;
+                LogError( ( "Only Qos 0 and 1 are supported over BLE." ) );
+                break;
+
+            /* Client tries to send a server to client only packet */
+            default:
+                status = MQTTBadParameter;
+                LogError( ( "A server to client only packet was sent. Check packet type or ensure Qos < 2." ) );
+                LogError( ( "Your packet type was %i", packetType ) );
+                break;
+        }
+    }
+    else
+    {
+        status = MQTTSendFailed;
+        LogError( ( "Serializer could not be initialized while sending data." ) );
+    }
+
+    IotBleMqtt_CleanupSerialize();
+
+    if( status != MQTTSuccess )
+    {
+        /* The user would have already seen a log for the previous error */
+        LogError( ( "No data was sent because of a previous error." ) );
+        MQTTBytesWritten = 0;
+    }
+    else
+    {
+        CBORbytesWritten = IotBleDataTransfer_Send( pContext->pChannel, bufToSend.pBuffer, pSize );
+    }
+
+    if( CBORbytesWritten == 0U )
+    {
+        LogError( ( "No data was sent because of the channel. Ensure that the channel is initialized and ready to send data." ) );
+        MQTTBytesWritten = 0;
+    }
+
+    return MQTTBytesWritten;
+}
+
+void IotBleMqttTransportAcceptData( const NetworkContext_t * pContext )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPacketInfo_t packet;
+    MQTTFixedBuffer_t pBuffer;
+
+    uint8_t sharedBuffer[ RECV_PACKET_BUFFER_SIZE ];
```

Comment:
Could be better if we can avoid this buffer allocation on stack as increasing RECV_PACKET_BUFFER_SIZE  can lead to stack overflow.  We could use the stream buffer allocated to serialize and send fixed bytes at a time.